subsys/trusted_storage/src/auth_crypt_backend.c

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <zephyr/device.h>
+#include <init.h>
+#include <sys/util.h>
+
+#include <stdio.h>
+#include <stdint.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+
+#include "settings_helpers.h"
+#include "auth_crypt_backend.h"
+
+/* AEAD IV suffix */
+#define TRUSTED_STORAGE_FILENAME_SUFFIX_IV		".iv"
+
+/*
+ * AEAD based Authenticated Encrypted trust implementation
+ *
+ * Actual implementation uses:
+ * - Hexadecimal UID imported as KEY
+ * - UID+Flags+Size as additional parameter
+ * - IV is an incrementing number at each encryption
+ * - TAG is left at the end of output data
+ */
+
+/* TODO Kconfig */
+#define AEAD_TAG_SIZE		16
+#define AEAD_IV_SIZE		12
```

Comment:
Why isn't the IV size 16, here?

Reply:
MbedTLS ChaChaPoly nonce is 12 bytes:
```
int mbedtls_chachapoly_starts( mbedtls_chachapoly_context *ctx,                 
                              const unsigned char nonce[12],                   
                               mbedtls_chachapoly_mode_t mode ); 
```

same for PSA AEAD with GCM:
```
#define PSA_AEAD_NONCE_LENGTH(key_type, alg) \                                  
    (PSA_BLOCK_CIPHER_BLOCK_LENGTH(key_type) == 16 ? \                          
          MBEDTLS_PSA_ALG_AEAD_EQUAL(alg, PSA_ALG_CCM) ? 13 : \                 
          MBEDTLS_PSA_ALG_AEAD_EQUAL(alg, PSA_ALG_GCM) ? 12 : \                 
          0 : \                                                                 
     (key_type) == PSA_KEY_TYPE_CHACHA20 && \                                   
          MBEDTLS_PSA_ALG_AEAD_EQUAL(alg, PSA_ALG_CHACHA20_POLY1305) ? 12 : \   
     0)       
```

Reply:
Ah... Yes. There is likely a collision in terms here dependent on how ChaCha20 is described compared to e.g. AES. 

Probably a more accurate term could be "nonce" for the thing that is 12 bytes long

The ChaCha20 algorithm input is explained as this (in RFC7539)

   o  A 32-bit initial counter.  This can be set to any number, but will
      usually be zero or one.  It makes sense to use one if we use the
      zero block for something else, such as generating a one-time
      authenticator key as part of an AEAD algorithm.

   o  A 96-bit nonce.  In some protocols, this is known as the
      Initialization Vector.
  
From my understanding the counter and nonce is "concatenated" as part of the inititialization state of the cipher. But it isn't exactly described as an initialization vector (concatenating counter + nonce)

Maybe comment that the IV-size corresponds to a nonce as described in ChaCha20. It would be neccesary to update this to 16 bytes for AES operations